src/App.jsx

```diff
@@ -0,0 +1,357 @@
+    const displayCanvas = displayCanvasRef.current;
+    const displayContext = displayCanvas.getContext('2d');
+    colorPicker();
+    context.fillRect(0, 0, canvas.width, canvas.height);
+    displayContext.drawImage(drawCanvas, 0, 0);
+    drawingHistory.push(canvas.toDataURL());
+    drawingNumber ++;
+  }
+
+  const resetCanvas = () => {
+    const drawCanvas = drawCanvasRef.current;
+    const context = drawCanvas.getContext('2d');
+    const displayCanvas = displayCanvasRef.current;
+    const displayContext = displayCanvas.getContext('2d');
+    drawingHistory.push(displayCanvas.toDataURL());
+    context.reset();
+    displayContext.reset();
+    drawingHistory.push(displayCanvas.toDataURL());
+    drawingNumber ++;
+    colorPicker();
+  }
+
+  const undo = () => {
+    const drawCanvas = drawCanvasRef.current;
+    const context = drawCanvas.getContext('2d');
+    const displayCanvas = displayCanvasRef.current;
+    const displayContext = displayCanvas.getContext('2d');
+
+    checkNumber = drawingNumber;
+    console.log(checkNumber, drawingNumber)
+    redoStates.push(displayCanvas.toDataURL());
+
+    if(drawingHistory.length > 0) {
+      context.clearRect(0, 0, canvas.width, canvas.height);
+      displayContext.clearRect(0, 0, canvas.width, canvas.height);
+      drawingHistory.pop();
+      const img = new Image();
+      img.src = drawingHistory[drawingHistory.length - 1];
+      img.addEventListener("load", () => {
+        displayContext.drawImage(img, 0, 0);
+        context.drawImage(img, 0, 0);
+      });
+    }
+  }
+
+  const redo = () => {
+    if(drawingNumber > checkNumber) {
+      redoStates = [];
+    }
+
+    const drawCanvas = drawCanvasRef.current;
+    const context = drawCanvas.getContext('2d');
+    const displayCanvas = displayCanvasRef.current;
+    const displayContext = displayCanvas.getContext('2d');
+
+    drawingHistory.push(canvas.toDataURL());
+
+    if (redoStates.length > 0) {
+      const img = new Image();
+      img.src = redoStates.pop();
+      img.addEventListener("load", () => {
+        context.clearRect(0, 0, canvas.width, canvas.height);
+        displayContext.clearRect(0, 0, canvas.width, canvas.height);
+        displayContext.drawImage(img, 0, 0);
+        context.drawImage(img, 0, 0);
+      });
+    };
+  };
+
+  window.addEventListener("keydown", (ev) => {
+    if (ev.key === "z" && ev.ctrlKey) {
+        undo();
+    };
+
+    if (ev.key === "z" && ev.metaKey) {
+        undo();
+    };
+  })
+
+  window.addEventListener("keydown", (ev) => {
+      if (ev.key === "y" && ev.ctrlKey) {
+          redo();
+      } else if (ev.key === "y" && ev.metaKey) {
+          redo();
+      };
+  });
+
+  window.addEventListener("mouseup", endDrawing);
+  window.addEventListener("touchend", endDrawing);
+
+  return (
+    <>
+      <div className="container">
+        <div className='toolbar'>
+          <Undo
+            undoHandler={undo}></Undo>
```

Comment:
Folosește Self-Closing-Tag când nu ai conținut. Sunt mai multe cazuri de genul.

```suggestion
          <Undo undoHandler={undo} />
```

Reply:
Am pus Self-Closing-Tag unde am considerat ca ar arata mai ok, dar probabil o sa mai modific cand incerc sa mai refac componentele.